livre/lib/pipeline/HistogramFilter.cpp

```diff
@@ -0,0 +1,178 @@
+/* Copyright (c) 2011-2015, EPFL/Blue Brain Project
+ *                     Ahmet Bilgili <ahmet.bilgili@epfl.ch>
+ *
+ * This file is part of Livre <https://github.com/BlueBrain/Livre>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <livre/lib/pipeline/HistogramFilter.h>
+#include <livre/lib/cache/HistogramCache.h>
+#include <livre/lib/cache/TextureDataCache.h>
+#include <livre/lib/cache/HistogramObject.h>
+
+#include <livre/core/data/DataSource.h>
+#include <livre/core/data/Histogram.h>
+#include <livre/core/render/Frustum.h>
+
+namespace livre
+{
+
+namespace
+{
+const float infinite = 100000.0f;
+}
+
+struct HistogramFilter::Impl
+{
+    Impl( HistogramCache& histogramCache )
+        : _histogramCache( histogramCache )
+        , _dataSource( histogramCache.getDataCache().getDataSource( ))
+        , _components( _dataSource.getVolumeInfo().compCount )
+    {}
+
+    bool isCenterInViewport( const Frustum& frustum,
+                             const Boxf& worldBox,
+                             const Vector3f& minNdc = Vector3f( -1.0f ),
+                             const Vector3f& maxNdc = Vector3f( 1.0f )) const
+    {
+        Vector4f center = worldBox.getCenter();
+        center[ 3 ] = 1.0;
+
+        Vector4f mvpCenter = frustum.getMVPMatrix() * center;
+        Vector3f mvpCenterHom = mvpCenter / mvpCenter[ 3 ];
+
+        const Boxf ndcCube( minNdc, maxNdc );
+        return ndcCube.isIn( mvpCenterHom );
+    }
+
+    void scaleNdcIfBorder( const Viewport& border, Vector3f& minBox, Vector3f& maxBox ) const
+    {
+        const bool isNegXBorder = border.x() == 0.0f;
+        const bool isPosXBorder = border.x() + border.z() == 1.0f;
```

Comment:
z is a bit confusing. I'd use [2] instead, which is as little informative as z, but at least it doesn't have an implied meaning.